File: tier1-web-client/js/jitter-buffer.js
/**
 * ✅ CRITICAL FIX: Client-Side Jitter Buffer for Voice
 *
 * Problem: Voice packets arrive out of order, causing audio glitches/gaps
 * Solution: Buffer audio packets and play them in order with adaptive delay
 *
 * Benefits:
 * - Smooth voice playback
 * - Handles network jitter
 * - Prevents audio stuttering
 * - Adaptive buffer size
 */
class JitterBuffer {
  constructor() {
    this.buffer = new Map(); // seq -> {timestamp, data, arrivalTime}
    this.nextExpectedSeq = 0;
    this.bufferDelayMs = 60; // Target 60ms buffer
    this.minBufferMs = 20;
    this.maxBufferMs = 200;
    this.lastPlayoutTime = 0;

    // Audio context for playback
    this.audioContext = null;
    this.nextPlayTime = 0;

    // Metrics
    this.packetsReceived = 0;
    this.packetsPlayed = 0;
    this.packetsLate = 0;
    this.packetsDropped = 0;
    this.totalJitter = 0;
    this.lastArrivalTime = 0;
    this.lastPacketTimestamp = 0;

    console.log("[JitterBuffer] ✅ Initialized (target: 60ms)");
  }

  /**
   * Initialize audio context
   */
  initAudioContext() {
    if (!this.audioContext) {
      this.audioContext = new (window.AudioContext ||
        window.webkitAudioContext)();
      this.nextPlayTime = this.audioContext.currentTime;
      console.log("[JitterBuffer] Audio context initialized");
    }
  }

  /**
   * Add audio packet to buffer
   * @param {number} seq - Sequence number
   * @param {number} timestamp - Packet timestamp (ms)
   * @param {ArrayBuffer} data - Audio data (Int16 PCM)
   */
  addPacket(seq, timestamp, data) {
    this.packetsReceived++;

    const currentTime = Date.now();

    // Calculate jitter
    if (this.lastArrivalTime > 0 && this.lastPacketTimestamp > 0) {
      const arrivalDelta = currentTime - this.lastArrivalTime;
      const timestampDelta = timestamp - this.lastPacketTimestamp;
      const jitter = Math.abs(arrivalDelta - timestampDelta);
      this.totalJitter += jitter;

      // Adapt buffer size based on jitter
      this.adaptBufferSize(jitter);
    }

    this.lastArrivalTime = currentTime;
    this.lastPacketTimestamp = timestamp;

    // Add to buffer
    this.buffer.set(seq, {
      timestamp: timestamp,
      data: data,
      arrivalTime: currentTime,
    });

    // Cleanup old packets (keep last 50 max)
    if (this.buffer.size > 50) {
      const oldestSeq = Math.min(...this.buffer.keys());
      this.buffer.delete(oldestSeq);
      this.packetsDropped++;
    }

    // Try to play buffered packets
    this.playBufferedPackets();
  }

  /**
   * Adapt buffer size based on observed jitter
   */
  adaptBufferSize(jitter) {
    if (jitter > this.bufferDelayMs * 0.8) {
      // High jitter, increase buffer
      this.bufferDelayMs = Math.min(this.maxBufferMs, this.bufferDelayMs + 10);
    } else if (jitter < this.bufferDelayMs * 0.3) {
      // Low jitter, decrease buffer for lower latency
      this.bufferDelayMs = Math.max(this.minBufferMs, this.bufferDelayMs - 5);
    }
  }

  /**
   * Play buffered packets that are ready
   */
  playBufferedPackets() {
    if (this.buffer.size === 0) {
      return;
    }

    this.initAudioContext();

    const currentTime = Date.now();

    // Find packets ready to play
    const readyPackets = [];

    for (const [seq, packet] of this.buffer.entries()) {
      const timeInBuffer = currentTime - packet.arrivalTime;

      if (timeInBuffer >= this.bufferDelayMs) {
        readyPackets.push({ seq, packet });
      }
    }

    // Sort by sequence number
    readyPackets.sort((a, b) => a.seq - b.seq);

    // Play ready packets
    for (const { seq, packet } of readyPackets) {
      this.playPacket(packet.data);
      this.buffer.delete(seq);
      this.packetsPlayed++;
      this.nextExpectedSeq = seq + 1;
    }
  }

  /**
   * Play audio packet
   * @param {ArrayBuffer} data - Audio data (Int16 PCM)
   */
  playPacket(data) {
    try {
      // Convert Int16 to Float32
      const int16Array = new Int16Array(data);
      const float32Array = new Float32Array(int16Array.length);

      for (let i = 0; i < int16Array.length; i++) {
        float32Array[i] = int16Array[i] / 32768.0;
      }

      // Create audio buffer
      const audioBuffer = this.audioContext.createBuffer(
        1,
        float32Array.length,
        this.audioContext.sampleRate
      );
      audioBuffer.getChannelData(0).set(float32Array);

      // Schedule playback
      const source = this.audioContext.createBufferSource();
      source.buffer = audioBuffer;
      source.connect(this.audioContext.destination);

      // Play at next scheduled time
      const playTime = Math.max(
        this.nextPlayTime,
        this.audioContext.currentTime
      );
      source.start(playTime);

      // Update next play time
      const duration = audioBuffer.duration;
      this.nextPlayTime = playTime + duration;
    } catch (error) {
      console.error("[JitterBuffer] Error playing packet:", error);
    }
  }

  /**
   * Get buffer status
   */
  getStatus() {
    return {
      bufferSize: this.buffer.size,
      bufferDelayMs: this.bufferDelayMs,
      packetsReceived: this.packetsReceived,
      packetsPlayed: this.packetsPlayed,
      packetsLate: this.packetsLate,
      packetsDropped: this.packetsDropped,
      averageJitter:
        this.packetsReceived > 1
          ? this.totalJitter / (this.packetsReceived - 1)
          : 0,
    };
  }

  /**
   * Print statistics
   */
  printStats() {
    const status = this.getStatus();
    console.log("\n=== Jitter Buffer Statistics ===");
    console.log(`Buffer Size: ${status.bufferSize} packets`);
    console.log(`Buffer Delay: ${status.bufferDelayMs} ms`);
    console.log(`Packets Received: ${status.packetsReceived}`);
    console.log(`Packets Played: ${status.packetsPlayed}`);
    console.log(`Packets Late: ${status.packetsLate}`);
    console.log(`Packets Dropped: ${status.packetsDropped}`);
    console.log(`Average Jitter: ${status.averageJitter.toFixed(2)} ms`);
    console.log("================================\n");
  }

  /**
   * Clear buffer
   */
  clear() {
    this.buffer.clear();
    this.nextExpectedSeq = 0;
    this.lastPlayoutTime = 0;

    if (this.audioContext) {
      this.nextPlayTime = this.audioContext.currentTime;
    }
  }

  /**
   * Cleanup
   */
  cleanup() {
    this.clear();

    if (this.audioContext && this.audioContext.state !== "closed") {
      this.audioContext.close();
      this.audioContext = null;
    }
  }
}

File: tier3-core-server/src/streaming/protocol/JitterBuffer.java
package streaming.protocol;

import java.util.concurrent.ConcurrentSkipListMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * ✅ CRITICAL FIX: Jitter Buffer for Voice Packets
 * 
 * Problem: Voice packets arrive out of order, causing audio glitches
 * Solution: Buffer packets and play them in order with adaptive delay
 * 
 * Benefits:
 * - Smooth voice playback
 * - Handles network jitter (variable latency)
 * - Adaptive buffer size based on jitter
 * - Prevents audio gaps and stuttering
 * 
 * Usage: Server-side buffering before sending to clients
 */
public class JitterBuffer {
    private final String clientId;
    private final ConcurrentSkipListMap<Long, VoicePacket> buffer;

    // Configuration
    private static final int MIN_BUFFER_MS = 20; // Minimum 20ms buffer
    private static final int MAX_BUFFER_MS = 200; // Maximum 200ms buffer
    private static final int TARGET_BUFFER_MS = 60; // Target 60ms buffer

    // State
    private long nextExpectedSeq = 0;
    private long bufferDelayMs = TARGET_BUFFER_MS;
    private long lastPlayoutTime = 0;

    // Metrics
    private final AtomicInteger packetsReceived = new AtomicInteger(0);
    private final AtomicInteger packetsPlayed = new AtomicInteger(0);
    private final AtomicInteger packetsLate = new AtomicInteger(0);
    private final AtomicInteger packetsDropped = new AtomicInteger(0);
    private final AtomicLong totalJitter = new AtomicLong(0);

    // Jitter calculation
    private long lastArrivalTime = 0;
    private long lastPacketTimestamp = 0;

    public JitterBuffer(String clientId) {
        this.clientId = clientId;
        this.buffer = new ConcurrentSkipListMap<>();
        this.lastPlayoutTime = System.currentTimeMillis();
    }

    /**
     * Add packet to jitter buffer
     * 
     * @param seq       Sequence number
     * @param timestamp Packet timestamp (milliseconds)
     * @param data      Audio data
     */
    public void addPacket(long seq, long timestamp, byte[] data) {
        packetsReceived.incrementAndGet();

        // Calculate jitter
        long currentTime = System.currentTimeMillis();
        if (lastArrivalTime > 0 && lastPacketTimestamp > 0) {
            long arrivalDelta = currentTime - lastArrivalTime;
            long timestampDelta = timestamp - lastPacketTimestamp;
            long jitter = Math.abs(arrivalDelta - timestampDelta);
            totalJitter.addAndGet(jitter);

            // Adapt buffer size based on jitter
            adaptBufferSize(jitter);
        }
        lastArrivalTime = currentTime;
        lastPacketTimestamp = timestamp;

        // Add to buffer
        VoicePacket packet = new VoicePacket(seq, timestamp, data, currentTime);
        buffer.put(seq, packet);

        // Cleanup old packets (older than 1 second)
        buffer.headMap(seq - 50).clear(); // Keep last 50 packets max
    }

    /**
     * Get next packet ready for playout
     * Returns null if no packet is ready yet (waiting for jitter buffer)
     */
    public byte[] getNextPacket() {
        if (buffer.isEmpty()) {
            return null;
        }

        long currentTime = System.currentTimeMillis();

        // Check if enough time has passed since last playout
        long timeSinceLastPlayout = currentTime - lastPlayoutTime;
        if (timeSinceLastPlayout < 20) { // Minimum 20ms between packets
            return null;
        }

        // Get oldest packet in buffer
        Long firstSeq = buffer.firstKey();
        if (firstSeq == null) {
            return null;
        }

        VoicePacket packet = buffer.get(firstSeq);

        // Check if packet has been buffered long enough
        long timeInBuffer = currentTime - packet.arrivalTime;
        if (timeInBuffer < bufferDelayMs) {
            // Not ready yet, wait for jitter buffer to fill
            return null;
        }

        // Remove and return packet
        buffer.remove(firstSeq);
        packetsPlayed.incrementAndGet();
        lastPlayoutTime = currentTime;
        nextExpectedSeq = firstSeq + 1;

        return packet.data;
    }

    /**
     * Adapt buffer size based on observed jitter
     */
    private void adaptBufferSize(long jitter) {
        if (jitter > bufferDelayMs * 0.8) {
            // High jitter, increase buffer
            bufferDelayMs = Math.min(MAX_BUFFER_MS, bufferDelayMs + 10);
        } else if (jitter < bufferDelayMs * 0.3) {
            // Low jitter, decrease buffer for lower latency
            bufferDelayMs = Math.max(MIN_BUFFER_MS, bufferDelayMs - 5);
        }
    }

    /**
     * Check if buffer has packets ready
     */
    public boolean hasReadyPackets() {
        if (buffer.isEmpty()) {
            return false;
        }

        long currentTime = System.currentTimeMillis();
        Long firstSeq = buffer.firstKey();
        if (firstSeq == null) {
            return false;
        }

        VoicePacket packet = buffer.get(firstSeq);
        long timeInBuffer = currentTime - packet.arrivalTime;

        return timeInBuffer >= bufferDelayMs;
    }

    /**
     * Get buffer size (number of packets)
     */
    public int getBufferSize() {
        return buffer.size();
    }

    /**
     * Get current buffer delay in milliseconds
     */
    public long getBufferDelayMs() {
        return bufferDelayMs;
    }

    /**
     * Get average jitter
     */
    public double getAverageJitter() {
        int received = packetsReceived.get();
        if (received <= 1) {
            return 0.0;
        }
        return (double) totalJitter.get() / (received - 1);
    }

    /**
     * Get metrics
     */
    public JitterMetrics getMetrics() {
        return new JitterMetrics(
                packetsReceived.get(),
                packetsPlayed.get(),
                packetsLate.get(),
                packetsDropped.get(),
                buffer.size(),
                bufferDelayMs,
                getAverageJitter());
    }

    /**
     * Print statistics
     */
    public void printStats() {
        System.out.println("\n=== Jitter Buffer Stats [" + clientId + "] ===");
        System.out.println("Packets Received: " + packetsReceived.get());
        System.out.println("Packets Played: " + packetsPlayed.get());
        System.out.println("Packets Late: " + packetsLate.get());
        System.out.println("Packets Dropped: " + packetsDropped.get());
        System.out.println("Current Buffer Size: " + buffer.size());
        System.out.println("Buffer Delay: " + bufferDelayMs + " ms");
        System.out.printf("Average Jitter: %.2f ms%n", getAverageJitter());
        System.out.println("===========================================\n");
    }

    /**
     * Clear buffer
     */
    public void clear() {
        buffer.clear();
        nextExpectedSeq = 0;
        lastPlayoutTime = System.currentTimeMillis();
    }

    /**
     * Voice packet wrapper
     */
    private static class VoicePacket {
        final long seq;
        final long timestamp;
        final byte[] data;
        final long arrivalTime;

        VoicePacket(long seq, long timestamp, byte[] data, long arrivalTime) {
            this.seq = seq;
            this.timestamp = timestamp;
            this.data = data;
            this.arrivalTime = arrivalTime;
        }
    }

    /**
     * Metrics data class
     */
    public static class JitterMetrics {
        public final int packetsReceived;
        public final int packetsPlayed;
        public final int packetsLate;
        public final int packetsDropped;
        public final int currentBufferSize;
        public final long bufferDelayMs;
        public final double averageJitter;

        JitterMetrics(int packetsReceived, int packetsPlayed, int packetsLate,
                int packetsDropped, int currentBufferSize, long bufferDelayMs,
                double averageJitter) {
            this.packetsReceived = packetsReceived;
            this.packetsPlayed = packetsPlayed;
            this.packetsLate = packetsLate;
            this.packetsDropped = packetsDropped;
            this.currentBufferSize = currentBufferSize;
            this.bufferDelayMs = bufferDelayMs;
            this.averageJitter = averageJitter;
        }
    }
}
